text/0000-release-channels.md

```diff
@@ -0,0 +1,70 @@
+
+- Feature Name: release\_channels
+- Start Date: 2018-06-30
+- RFC PR:
+- Bundler Issue:
+
+# Summary
+
+Bundler release channels allow users to opt in to the level of release stability that they want. The “canary” channel has a release from every commit that builds green. The default channel has regular releases when features and bugfixes are ready. The “stable” channel only has releases that have fixes for any issues found by the default channel. When possible, bugfixes are backported from the default channel to the stable channel.
+
+# Motivation
+
+As the developers of Bundler, we want to be able to continue to develop and revise Bundler into a better version of itself. As users of Bundler, we want the safest, most stable, least problematic version of Bundler at any given time.
+
+Release channels are a way for the Bundler team to officially bless and support versions that have proven themselves to be stable. It also adds an easy and straightforward way to try the latest upcoming changes to Bundler at any time, increasing the chances that bugs can be found and fixed.
+
+With the stable channel, this RFC aims to actively support the usecases of users deploying Ruby code to important production servers. With the canary channel, this RFC aims to actively support power users, Bundler contributors, early adopters, and anyone who wants to participate in the future of Bundler.
+
+Hopefully, release channels will allow Bundler to change and move forward more easily, while at the same time providing even more stability as an option for users who care most about that.
+
+# Guide-level explanation
+
+To use the default release channel, install the Bundler gem and use it as usual. For example, you install Bundler by running `gem install bundler`, and you install gems by running `bundle install`. The default release of Bundler will let you know anytime there is a newer default release available, but will not install it for you.
+
+To use the stable release channel, instead install with `gem install bundler-stable`. Install gems as usually, by running `bundle install`. The stable version of Bundler will let you know anytime there is a newer stable version available, but will not install it for you. You can identify stable releases of Bundler by running `bundle install --version` and looking for `(stable)` in the output.
+
+To use the canary release channel, instead install with `gem install bundler-canary`. Install gems as usually, by running `bundle install`. The canary version of Bundler will continuously keep itself up to date, installing the latest canary version (if you don’t already have it) during any `bundle install`. You can identify canary versions of Bundler by running `bundle --version`, and checking the output for `(canary)`.
+
+# Reference-level explanation
+
+Default releases are created using today’s release process, running `rake release`. Maybe default releases could be the same as canary releases, but with different feature flags?
```

Comment:
It is not clear to me why `gem install bundler` don't install what you are calling the stable channel and `gem install bundler-canary` install what you are calling the default channel, care to expand the reasoning why the we have the tree channels instead of only two (canary and stable)?

Reply:
Sure, I can take another stab at it. Most users want a "pretty stable" version, with "pretty recent" bugfixes. The current releases try to do that, striking a midpoint between release frequency and release stability.

If we made `gem install bundler` install the stable channel, nothing would change from how things are today... most users would not run into bugs until the wide release, many users would run into bugs on each release, and Heroku would continue to not upgrade for months or years in an effort to stay away from new issues.

This RFC is an attempt to offer a different option, where opting in to the canary channel can either be a way to get the latest build from the master branch _or_ be a way to always use the most recent version of Bundler that has passed all the tests in order to provide feedback.

The regular release channel then becomes the second tier of frequency/stability. We'll continue to fix known bugs before releasing, and hopefully more bugs will be caught thanks to the help of canary users.

Finally, the stable release channel is a way for most or all edge cases to be caught before we ship a release suitable for Heroku or other PaaSes to use—after a huge number of regular users have tried the release and we have ironed out any additional bugs.

In practice, this three-tier system is what we already have. Cutting edge users install Bundler from git, regular users run the latest regular release, and Heroku only installs a single carefully selected bugfix release once every 2-3 years. This RFC is about making it easier for all of those groups to participate, give us feedback, and install new releases with less stress.

Reply:
That makes sense and I like the plan. How would the `BUNDLED_WITH` option works for cases were users have the regular channel and their machine but deploy to an infra that uses the stable channel. Would it show a warning? I believe that warning will not be actionable for the developer, so maybe it will be perceived as noise and ignored most of the time?

Reply:
Yeah, I'm not sure what would be best. We can either leave the warning the way it is today, and it will be noise/ignored, or we can change the warning to check for newer versions only on the specific channel that is installed. I think either one of those would probably be fine?

Reply:
Checking new versions of the specific channel would be ideal, but if that requires a lot more code we can leave the warning in the way it is today.